src/compiler/cse.rs

```diff
@@ -0,0 +1,452 @@
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    marker::PhantomData,
+};
+
+use crate::{
+    field::Field,
+    poly::{
+        self,
+        cse::{create_common_ses_signal, replace_expr},
+        Expr, HashResult, VarAssignments,
+    },
+    sbpir::{
+        query::Queriable, sbpir_machine::SBPIRMachine, ForwardSignal, InternalSignal, StepType,
+    },
+    wit_gen::NullTraceGenerator,
+};
+
+#[derive(Clone, Debug)]
+pub struct CseConfig {
```

Comment:
@rutefig we are commenting all public interfaces now, but perhaps this can be just pub(crate), many things here perhaps are just pub(crate) because they are only called internelly from the compiler.